cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,302 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+// Functor to identify unique elements in a sorted order table/column
+template <bool has_nulls, typename ReturnType = bool>
+struct unique_comparator {
+  unique_comparator(table_device_view device_table,
+                    size_type const *sorted_order)
+      : comparator(device_table, device_table, true),
+        permute(sorted_order) {}
+  __device__ ReturnType operator()(size_type index) const noexcept {
+    return index == 0 || not comparator(permute[index], permute[index - 1]);
+  };
+
+ private:
+  row_equality_comparator<has_nulls> comparator;
+  size_type const *permute;
+};
+
+// Assign rank from 1 to n unique values. Equal values get same rank value.
+rmm::device_vector<size_type> sorted_dense_rank(column_view input_col,
+                                                column_view sorted_order_view,
+                                                cudaStream_t stream) {
+  auto device_table = table_device_view::create(table_view{{input_col}}, stream);
+  auto const input_size = input_col.size();
+  rmm::device_vector<size_type> dense_rank_sorted(input_size);
+  if (input_col.has_nulls()) {
+    auto conv = unique_comparator<true, size_type>(
+        *device_table, sorted_order_view.data<size_type>());
+    auto unique_it = thrust::make_transform_iterator(
+        thrust::make_counting_iterator<size_type>(0), conv);
+
+    thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           unique_it,
+                           unique_it + input_size,
+                           dense_rank_sorted.data().get());
+  } else {
+    auto conv = unique_comparator<false, size_type>(
+        *device_table, sorted_order_view.data<size_type>());
+    auto unique_it = thrust::make_transform_iterator(
+        thrust::make_counting_iterator<size_type>(0), conv);
+
+    thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           unique_it,
+                           unique_it + input_size,
+                           dense_rank_sorted.data().get());
+  }
```

Comment:
This can be simplified (because `thrust::make_transform_it..` and `thrust::inclusive_scan` are common.
```cpp
auto conv = [&] {
    if (input_col.has_nulls())   
        return unique_comparator<true, size_type>(
            *device_table, sorted_order_view.data<size_type>());
    else
        return unique_comparator<false, size_type>(
            *device_table, sorted_order_view.data<size_type>());
} ();

auto unique_it = thrust::make_transform_iterator(
    thrust::make_counting_iterator<size_type>(0), conv);

thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
                       unique_it,
                       unique_it + input_size,
                       dense_rank_sorted.data().get());
}
```
This is where we wish we had a `constexpr` ternary operator :(

Reply:
This won't work. `unique_comparator<true, size_type>` is different type from `unique_comparator<false, size_type>`.

Reply:
Right you are. Is it possible to still combine the `thrust::inclusive_scan`s?
```cpp
auto unique_it = [&] {
    if (input_col.has_nulls()) {   
        auto conv = unique_comparator<true, size_type>(
            *device_table, sorted_order_view.data<size_type>());
        return thrust::make_transform_iterator(
            thrust::make_counting_iterator<size_type>(0), conv);
    } else {
        return unique_comparator<false, size_type>(
            *device_table, sorted_order_view.data<size_type>());
        return thrust::make_transform_iterator(
            thrust::make_counting_iterator<size_type>(0), conv);
    }
} ();

thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
                       unique_it,
                       unique_it + input_size,
                       dense_rank_sorted.data().get());
}
```

Reply:
Tried this as well. It's same issue - different types.

Reply:
Ok.

Reply:
I thought this was the case, but @codereport got me excited that I was missing something. This is one of the most annoying patterns in our codebase to me.